Phoenix bc4 feature bjs2 6778

Murzin34 opened the pull request.

Murzin34 committed:
initial

Murzin34 committed:
change some logic

Murzin34 committed:
small logic fix

Murzin34 committed:
add test

Murzin34 committed:
Внес изменения в модель Resource и таблицу resource_allowed_roles

NadirCianid commented on src/main/resources/db/changelog/changeset/project_V014_alter_table_resource_allowed_roles.sql:
```diff
@@ -0,0 +1,5 @@
+alter table resource_allowed_roles
+    rename column role_id to role;
+
+alter table resource_allowed_roles
+    alter column role type varchar using role::varchar;
```
Во-первых, ты не включил этот файл в db.changelog, так что он считай не учитывается при создании и обновлении БД.

Во-вторых, такой таблицы нет.

В-третьих, я не понимаю смысл этих изменений. Возможно была бы таблица, я бы понял...

Murzin34 replied:
Добавил в db.changelog-master.yam
По поводу отличий наших бд будем разбираться.

NadirCianid commented on src/main/java/faang/school/projectservice/config/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,47 @@
+@Configuration
+@RequiredArgsConstructor
+public class AmazonS3Config {
+    @Value("${services.s3.endpoint}")
+    private String endpoint;
+    @Value("${services.s3.accessKey}")
+    private String accessKey;
+    @Value("${services.s3.secretKey}")
+    private String secretKey;
+    @Value("${services.s3.bucketName}")
+    private String bucketName;
```
Почитай про аннотацию @ConfigurationProperties, на в данном случае отлично подходит

Murzin34 replied:
Добавил `@ConfigurationProperties(prefix = "services.s3")`

NadirCianid commented on src/main/java/faang/school/projectservice/config/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Bean
+    public AmazonS3 amazonS3() {
+        AWSStaticCredentialsProvider awsStaticCredentialsProvider =
+                new AWSStaticCredentialsProvider(
+                        new BasicAWSCredentials(accessKey, secretKey)
+                );
+        AwsClientBuilder.EndpointConfiguration endpointConfiguration =
+                new AwsClientBuilder.EndpointConfiguration(endpoint, Regions.DEFAULT_REGION.getName());
```
Мне кажется создавать отдельные переменные для этих объектов немного избыточно и можно их прямо в билдере создать

Murzin34 replied:
Поменял реализацию, удалил отдельные переменные.

NadirCianid commented on src/main/java/faang/school/projectservice/config/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,47 @@
+                .withPathStyleAccessEnabled(true)
+                .build();
+
+        if (!s3client.doesBucketExistV2(bucketName)) {
+            s3client.createBucket(bucketName);
+        }
+        return s3client;
```
было бы неплохо добавить логирование сюда

Murzin34 replied:
Добавил логирование в виде `log.info("Connected to AmazonS3.")`

NadirCianid commented on src/main/java/faang/school/projectservice/service/s3/AmazonS3Service.java:
```diff
@@ -0,0 +1,51 @@
+            throw new S3Exception(e.getMessage());
+        }
+
+        return new Resource();
```
Почему возвращаешь пустой объект?

Murzin34 replied:
AmazonS3Service только получаю объект. Новый объект передаю в ResourceService, где и происходит полноценная работа с ним.

Murzin34 replied:
Реализацию удалил, использую уже существующую в мастере

NadirCianid commented on src/main/java/faang/school/projectservice/exception/S3Exception.java:
```diff
@@ -0,0 +1,7 @@
+package faang.school.projectservice.exception;
+
+public class S3Exception extends RuntimeException {
+    public S3Exception(String message) {
+        super(message);
+    }
+}
```
Ошибки создал (эта и еще одна), а обработчики не добавил. Плохо

Murzin34 replied:
Поправил

NadirCianid commented on src/main/java/faang/school/projectservice/controller/resource/ResourceController.java:
```diff
@@ -0,0 +1,37 @@
+    @PostMapping(value = "/{userId}/files")
+    @ResponseStatus(HttpStatus.CREATED)
+    public ResourceDto saveFile(@PathVariable @Positive Long userId,
+                                @NotNull @RequestBody MultipartFile file) {
```
Каким образом параметр помеченный аннотацией @RequestBody может быть null?

Murzin34 replied:
Убрал  `@NotNull`

NadirCianid commented on src/main/java/faang/school/projectservice/controller/resource/ResourceController.java:
```diff
@@ -0,0 +1,37 @@
+public class ResourceController {
+    private final ResourceService resourceService;
+
+    @PostMapping(value = "/{userId}/files")
```
Насколько я понял, в задании речь о файлах проекта, а не о файлах пользователя

NadirCianid replied:
Если уж ты решил это делать через членов команды (TeamMember), то было бы логичнее указывать именно TeamMemberId

Murzin34 replied:
Текст задания:
Участники проекта должны иметь возможность загружать файлы в проект, которые будут доступны другим участникам.

На мой взгляд это условие соблюдено. Можно, правда, создать дополнительный функционал именно под `@PostMapping(value = "/{projectId}/files")`

Ссылки и имена переменных поправил.

NadirCianid commented on src/main/java/faang/school/projectservice/controller/resource/ResourceController.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @GetMapping("/{userId}/files/{resourceId}")
+    public InputStreamResource getFile(@PathVariable @Positive Long userId,
```
Почему именно этот тип возвращаем ?

Murzin34 replied:
т.к. в AmazonS3Service использую S3ObjectInputStream. 
Можно использовать ByteArrayResource, но придется преобразовывать поток в байтовый массив. Вроде как ByteArrayResource выгодно использовать при придаче небольших файлов, а для крупных InputStreamResource.

NadirCianid commented on src/main/java/faang/school/projectservice/dto/resource/ResourceDto.java:
```diff
@@ -0,0 +1,32 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class ResourceDto {
+    private Long id;
+    private String key;
+    private String name;
+    private BigInteger size;
+    private Set<TeamRole> allowedRoles;
+    private ResourceType type;
+    private ResourceStatus status;
+    private Long createdById;
+    private Long updatedById;
+    private LocalDateTime createdAt;
+    private LocalDateTime updatedAt;
+    private Long projectId;
```
Не помешает навесить валидационные аннотации

Murzin34 replied:
Буду использовать аннотации из мастера

NadirCianid commented on src/main/java/faang/school/projectservice/jpa/TeamMemberJpaRepository.java:
```diff
@@ -11,9 +10,9 @@
-        "WHERE tm.userId = :userId " +
-        "AND t.project.id = :projectId"
+            "SELECT tm FROM TeamMember tm JOIN tm.team t " +
+                    "WHERE tm.userId = :userId " +
+                    "AND t.project.id = :projectId"
```
Зачем?

Murzin34 replied:
Идея так поправила. Создавал в этом репозитории метод который потом удалил.

NadirCianid commented on src/main/java/faang/school/projectservice/mapper/ResourceMapper.java:
```diff
@@ -0,0 +1,20 @@
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface ResourceMapper {
+    @Mapping(source = "createdById", target = "createdBy.id")
+    @Mapping(source = "updatedById", target = "updatedBy.id")
+    @Mapping(source = "projectId", target = "project.id")
```
Не лучшая идея. В итоге у тебя будут пустые объекты только с id'шниками. Нужно отдельно выцеплять из базы данные сущности и сетить их

Murzin34 replied:
Ровно такая же реализация мапера в мастере.

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -0,0 +1,105 @@
+    private final ResourceRepository resourceRepository;
+    private final ProjectRepository projectRepository;
+    private final ResourceMapper resourceMapper;
+    private final TeamMemberJpaRepository teamMemberRepository;
```
У тебя есть специальный класс TeamMemberRepository. Его надо использовать. В нем уже есть метод поиска члена команды с обработкой Optional

Murzin34 replied:
Поправил

NadirCianid commented on src/main/java/faang/school/projectservice/validator/resource/ResourceValidator.java:
```diff
@@ -0,0 +1,42 @@
+    public void validateMaxFreeStorageSize(Project project, Long fileLength) {
+        BigInteger newLength = project.getStorageSize().add(BigInteger.valueOf(fileLength));
+        if (newLength.compareTo(maxFreeStorageSize) > 0) {
+            throw new DataValidationException("Project storage is full");
```
Тут скорее ошибка в том, что не хватает места, чтобы добавить именно этот файл. Может тебе пришел файл весом в 100 Гб, а в хранилище проекта пусто. Это же не значит, что места нет. Это значит, что файл не помещается

Murzin34 replied:
Данный валидатор удалил, использую существующий вариант из мастера.

NadirCianid commented on src/main/java/faang/school/projectservice/validator/resource/ResourceValidator.java:
```diff
@@ -0,0 +1,42 @@
+@Component
+public class ResourceValidator {
+    @Value("${services.s3.maxFreeStorageSize}")
+    private BigInteger maxFreeStorageSize;
```
Я считаю, что данный параметр не относится именно к S3. Это параметр проекта. А в той группе, что ты указал лежат параметры для настройки подключения к серверу S3. Может найдем более подходящее место под этот параметр?

Murzin34 replied:
Данное поле удалил, использую существующий вариант из мастера.

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -0,0 +1,105 @@
+        Project project = teamMember.getTeam().getProject();
+
+        resourceValidator.validateMaxFreeStorageSize(project, file.getSize());
+        BigInteger newStorageSize = project.getStorageSize().add(BigInteger.valueOf(file.getSize()));
```
Дублируешь действие из валидатора

Murzin34 replied:
Данную реализацию добавления ресурса удалил, использую существующую из мастера.

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -0,0 +1,105 @@
+        BigInteger newStorageSize = project.getStorageSize().add(BigInteger.valueOf(file.getSize()));
+
+        String folder = project.getName() + project.getId();
+        String key = String.format("%s/%d%s", folder, System.currentTimeMillis(), file.getOriginalFilename());
```
Возможно это вкусовщина, но мне кажется странным использовать слеш в названии файла. Этот символ используется в основном для отделения каталогов (папок в ФС). Может лучше "_"?

Murzin34 replied:
Данную реализацию добавления ресурса удалил, использую существующую из мастера.

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -0,0 +1,105 @@
+        resourceValidator.validateMaxFreeStorageSize(project, file.getSize());
+        BigInteger newStorageSize = project.getStorageSize().add(BigInteger.valueOf(file.getSize()));
+
+        String folder = project.getName() + project.getId();
+        String key = String.format("%s/%d%s", folder, System.currentTimeMillis(), file.getOriginalFilename());
```
Я бы вынес это в отдельный метод, чтобы разгрузить текущий. Уж больно он пухлый получился

Murzin34 replied:
Данную реализацию добавления ресурса удалил, использую существующую из мастера.

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -0,0 +1,105 @@
+        String key = String.format("%s/%d%s", folder, System.currentTimeMillis(), file.getOriginalFilename());
+        Resource resource = amazonS3Service.uploadFile(key, file);
+
+        resource.setKey(key);
+        resource.setName(file.getOriginalFilename());
+        resource.setSize(BigInteger.valueOf(file.getSize()));
+        resource.setType(ResourceType.getResourceType(file.getContentType()));
+        resource.setAllowedRoles(new HashSet<>(teamMember.getRoles()));
+        resource.setStatus(ResourceStatus.ACTIVE);
+        resource.setCreatedBy(teamMember);
+        resource.setUpdatedBy(teamMember);
+        resource.setProject(project);
```
Вот это точно стоит выделить в отдельный метод

Murzin34 replied:
Данную реализацию добавления ресурса удалил, использую существующую из мастера.

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -0,0 +1,105 @@
+        String key = String.format("%s/%d%s", folder, System.currentTimeMillis(), file.getOriginalFilename());
+        Resource resource = amazonS3Service.uploadFile(key, file);
+
+        resource.setKey(key);
+        resource.setName(file.getOriginalFilename());
+        resource.setSize(BigInteger.valueOf(file.getSize()));
+        resource.setType(ResourceType.getResourceType(file.getContentType()));
```
данный блок можно было заполнить в amazonS3Service в билдере

Murzin34 replied:
Данную реализацию добавления ресурса удалил, использую существующую из мастера.

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -0,0 +1,105 @@
+        resourceRepository.save(resource);
+    }
+
+    private Resource getResource(Long resourceId) {
+        return resourceRepository.findById(resourceId)
+                .orElseThrow(() -> new NotFoundException("Not found resource with this id: " + resourceId));
+    }
```
Этот метод отлично подходит для класса-обертки репозитория. Посмотри на аналогичный в TeamMemberRepository

Murzin34 replied:
Поправил

NadirCianid commented on src/main/java/faang/school/projectservice/validator/resource/ResourceValidator.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public void validateDownloadFilePermission(TeamMember teamMember, Resource resource) {
+        if (!resource.getProject().equals(teamMember.getTeam().getProject())
+                && !teamMember.getRoles().contains(TeamRole.MANAGER)
+                && !teamMember.getRoles().contains(TeamRole.OWNER)) {
```
Тут не лучше использовать оператор ИЛИ ?

Между первым условием и 2&3*

NadirCianid replied:
А если файл попытается скачать разработчик из этого проекта? Тоже откажешь. 

При условии, что файл загрузил тоже разработчик*

Murzin34 replied:
Поправил реализацию валидатора

NadirCianid commented on src/test/java/faang/school/projectservice/service/resource/ResourceServiceTest.java:
```diff
@@ -0,0 +1,158 @@
+
+    @BeforeEach
+    void init() {
+        file = new MockMultipartFile("name", "original_name", "TEXT", new byte[10000]);
+        project = getProject();
+        teamMember = getTeamMember();
+        resource = getResource();
+        resourceDto = getResourceDto();
```
Это лучше вынеси в отдельный класс. Какой-нибудь TestData

Murzin34 replied:
Добавил ResourceServiceTestData

NadirCianid commented on src/test/java/faang/school/projectservice/service/resource/ResourceServiceTest.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Test
+    void testSaveFileSuccess() {
```
позитивные тесты вижу, а негативные нет

Murzin34 replied:
Поставил /TODO

NadirCianid commented on src/test/java/faang/school/projectservice/service/resource/ResourceServiceTest.java:
```diff
@@ -0,0 +1,158 @@
+    void testSaveFileSuccess() {
+        when(teamMemberRepository.findById(userId)).thenReturn(Optional.of(teamMember));
+        when(amazonS3Service.uploadFile(anyString(), any(MultipartFile.class))).thenReturn(resource);
+        when(resourceRepository.save(resource)).thenReturn(resource);
+        when(projectRepository.save(project)).thenReturn(project);
```
Зачем мокаешь это поведение? Ты же не используешь их результат. Лучше проверь их вызов, с помощью, verify

Murzin34 replied:
Поправил

NadirCianid commented on src/test/java/faang/school/projectservice/service/resource/ResourceServiceTest.java:
```diff
@@ -0,0 +1,158 @@
+        InputStreamResource expected = new InputStreamResource(InputStream.nullInputStream());
+        when(resourceRepository.findById(resourceId)).thenReturn(Optional.of(resource));
+        when(teamMemberRepository.findById(userId)).thenReturn(Optional.of(teamMember));
+        when(amazonS3Service.downloadFile(resource.getKey())).thenReturn(expected);
+
+        InputStreamResource actual = resourceService.getFile(userId, resourceId);
+
+        assertNotNull(actual);
+        assertEquals(expected, actual);
```
Ты не должен здесь проверять, что тебе вернул метод, а просто проверить факт его вызова. Так как сам метод ты должен проверить в отдельном тесте

Murzin34 replied:
Создал новый тест только под проверку.

NadirCianid commented on src/test/java/faang/school/projectservice/service/resource/ResourceServiceTest.java:
```diff
@@ -0,0 +1,158 @@
+
+    @Test
+    void testDeleteFileSuccess() {
+        when(resourceRepository.findById(resourceId)).thenReturn(Optional.of(resource));
+        when(teamMemberRepository.findById(userId)).thenReturn(Optional.of(teamMember));
```
Чувствуешь, что повторяются этим выражения? Может есть способ это исправить? Например вложенные классы или сделать это в сетапе, если в каждом методе используешь

Murzin34 replied:
Поправил, перенес в инит

NadirCianid commented on src/test/java/faang/school/projectservice/service/resource/ResourceServiceTest.java:
```diff
@@ -0,0 +1,158 @@
+
+        resourceService.deleteFile(userId, resourceId);
+
+        assertNull(resource.getKey());
+        assertNull(resource.getSize());
+        assertEquals(ResourceStatus.DELETED, resource.getStatus());
+        assertEquals(teamMember, resource.getUpdatedBy());
+        assertEquals(expectedLength, project.getStorageSize());
```
Не проверил вызовы методов amazonS3,  репозиториев и валидаторов

Murzin34 replied:
Поправил

NadirCianid reviewed (changes requested):
У меня не работает функционал. Проблемы с бд. Проверь все и протестируй
![image](https://github.com/CorporationX/project_service/assets/101037555/9ff3b588-3198-4028-bced-aaeaad6766d3)


Также стоит добавить логи

Murzin34 committed:
some fix 1

Murzin34 committed:
попытка мерджа из мастера

Murzin34 committed:
some fix 2

Murzin34 committed:
закончил merge из мастера

Murzin34 committed:
фикс миграции, удалил свой AmazonS3Config и использовал S3BeanConfig из мастера чуть его изменив.
фикс моего функционала, но нужно совместить его с уже существующеми в мастере методами

Murzin34 committed:
Добавил методы в сервис, исправил валидатор, тесты

Murzin34 committed:
Добавил в обработчик NoAccessException, тесты для валидатора

NadirCianid commented on src/main/java/faang/school/projectservice/controller/resource/TeamMemberResourceController.java:
```diff
@@ -0,0 +1,57 @@
+@RequiredArgsConstructor
+@Validated
+@RequestMapping("/resources/teamMember")
+public class TeamMemberResourceController {
```
Не должно быть два контроллера ресурсов. Я думаю стоит использовать тот ,что из мастера, по-хорошему

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -42,41 +42,41 @@ public class ResourceService {
+
         Resource saved = resourceRepository.save(createResource(project, teamMember, putRequest, resourceDto));
-        project.addStorageSize(saved.getSize());
+        project.setStorageSize(saved.getSize());
```
Ты же добавляешь файл, а значит, что объем памяти проекта увеличивается на размер файла. Не понимаю, почему ты решил каждый раз устанавливать объем памяти проекта в размер загруженного ресурса

Murzin34 replied:
Косяк, поправил изначально используемый метод из Project.java

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -85,35 +85,64 @@ public ResourceDto uploadResource(
         ResourceDto resourceDto = uploadResource(projectId, teamMemberId, file, ProjectResourceType.PROJECT_COVER);
-        
+
         project.addStorageSize(resourceDto.getSize());
```
В методе `uploadResource` ты уже изменил размер storageSize (хоть и неправильно) здесь это действие лишнее

NadirCianid commented on src/main/java/faang/school/projectservice/service/resource/ResourceService.java:
```diff
@@ -85,35 +85,64 @@ public ResourceDto uploadResource(
-        
+
         project.addStorageSize(resourceDto.getSize());
         project.setCoverImageId(resourceDto.getKey());
```
Может это действие убрать в обработчик обложек проекта. Только из-за него создавать этот метод кажется мне странным

NadirCianid reviewed (dismissed):
В целом норм, но не понял, зачем ты оставил два контроллера

И изменение объема хранимых данных проекта исправь

NadirCianid commented on src/main/java/faang/school/projectservice/controller/resource/TeamMemberResourceController.java:
```diff
@@ -0,0 +1,57 @@
+
+    @GetMapping()
+    @ResponseStatus(HttpStatus.OK)
+    public InputStreamResource getResources(
```
![image](https://github.com/CorporationX/project_service/assets/101037555/230ba67d-28b9-4ef2-80c8-484713629da0)

Какая-то фигня

Murzin34 replied:
txt в Postman'e отображается корректно, а вот изображения нет. Ибо JSON.
Поправил. В Postman'e все также плохо выглядит, но через браузер файл скачивается корректно.

NadirCianid reviewed (commented):
Что-то не так с возвращаемым типом в методе получения ресурса

Murzin34 committed:
Фикс реализации Get запроса в TeamMemberResourceController,фикс отображения StorageSize в Project

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/controller/resource/TeamMemberResourceController.java:
```diff
@@ -0,0 +1,58 @@
+public class TeamMemberResourceController {
+    private final ResourceService resourceService;
+
+    @PostMapping()
```
~~()~~

Murzin34 replied:
Поправил

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/controller/resource/TeamMemberResourceController.java:
```diff
@@ -0,0 +1,58 @@
+        return resourceService.getResources(teamMemberId, resourceId);
+    }
+
+    @DeleteMapping()
```
~~()~~

Murzin34 replied:
Поправил

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/validator/resource/TeamMemberResourceValidator.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    public void validateDeleteFilePermission(TeamMember teamMember, Resource resource) {
+        if (!resource.getCreatedBy().equals(teamMember)
```
разбей на мелкие приватные методы

Murzin34 replied:
Разбил

Murzin34 committed:
Фикс согласно замечаниям, добавил DisplayName на методах TeamMemberResourceValidatorTest

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/validator/resource/TeamMemberResourceValidator.java:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    private static boolean notUploader(TeamMember teamMember, Resource resource) {
```
а зачем они статики? обычный приватные методы и лучше переназвать с is или has (isNotUploader, isNoManagerFromUploaderProject, isNotOwnerFromUploaderProject)

Murzin34 replied:
Так idea подсказала (Ctrl+Alt+M).
Статики убрал, методы переименовал.

Sabina-Vishnya reviewed (dismissed):
убери статики в валидаторе, переименуй методы и вливай

Murzin34 committed:
фикс методов в TeamMemberResourceValidator
